### PR TITLE
[Op][Ref][CPU] Extend Gather op with String support

### PR DESCRIPTION
**Pull request by mitruska**
### Details:
 - Extend Gather op with String support for data input 
 (agreed to do it within the same op version, existing reference implementation is common for GatherBase and ready to handle strings, ref changes are not needed, just evaluate supported types updated)
 - Enable fallback to the reference implementation from CPU plugin
 - Extend Template and CPU plugin operator tests with string validation (most of the changes are tests related)
### Tickets:
 - 130688

### CODE REVIEW

**Review by rkazants — COMMENTED** [review]
Please unskip layer test case for `LoopupTableFindV2` with string values:
https://github.com/openvinotoolkit/openvino/pull/22517/files#diff-f7cc5818966bc741f8fe189e5704ddbc4406c16db971b9696f34bc9e85886eedR64

**Comment by mitruska** [comment]
> Please unskip layer test case for `LoopupTableFindV2` with string values

@rkazants As we discussed, this test will be enabled with string support in Reshape and Concat operations

File: src/core/src/op/gather.cpp [hunk 1]
```diff
@@ -108,6 +108,16 @@ std::shared_ptr<Node> Gather::clone_with_new_inputs(const OutputVector& new_args
...
     return std::make_shared<Gather>(new_args.at(0), new_args.at(1), new_args.at(2), m_batch_dims);
 }
+
+bool Gather::has_evaluate() const {
+    OV_OP_SCOPE(v8_Gather_has_evaluate);
+    switch (get_output_element_type(0)) {
+    case element::string:
+        return true;
+    default:
+        return false;
```
Comment (rkazants) [inline src/core/src/op/gather.cpp:118]: Don't we need `evaluate` for other types except `string`?

**Review by rkazants — APPROVED** [review]

File: src/tests/functional/shared_test_classes/src/single_op/gather.cpp [hunk 2]
```diff
@@ -244,5 +244,77 @@ void Gather8withIndicesDataLayerTest::SetUp() {
...
+    result << "indices_shape=" << ov::test::utils::vec2str(indices_shape) << "_";
+    result << "netPRC=" << model_type.get_type_name() << "_";
+    result << "trgDev=" << device_name << "_";
+    result << "indicesData=" << ov::test::utils::vec2str(indicesData) << "_";
+
+    return result.str();
+}
+
+void GatherStringWithIndicesDataLayerTest::SetUp() {
+    gatherStringParamsTuple basicParams = GetParam();
```
Comment (t-jankowski) [inline src/tests/functional/shared_test_classes/src/single_op/gather.cpp:285]:
```suggestion
    const gatherStringParamsTuple& basicParams = GetParam();
```

File: src/tests/functional/shared_test_classes/src/single_op/gather.cpp [hunk 3]
```diff
@@ -244,5 +244,77 @@ void Gather8withIndicesDataLayerTest::SetUp() {
     function = std::make_shared<ov::Model>(result, ov::ParameterVector{param}, "gather");
 }
 
+// Gather String support
+std::string GatherStringWithIndicesDataLayerTest::getTestCaseName(const testing::TestParamInfo<gatherStringParamsTuple>& obj) {
+    gatherStringParamsTuple basicParams = obj.param;
```
Comment (t-jankowski) [inline src/tests/functional/shared_test_classes/src/single_op/gather.cpp:249]:
```suggestion
    const gatherStringParamsTuple& basicParams = obj.param;
```

File: src/tests/functional/shared_test_classes/src/single_op/gather.cpp [hunk 4]
```diff
@@ -244,5 +244,77 @@ void Gather8withIndicesDataLayerTest::SetUp() {
...
+    std::vector<int64_t> indicesData;
+    std::vector<std::string> str_data;
+
+    std::tuple<int, int> axis_batch_idx;
+    std::vector<int> indices;
+    ov::Shape indices_shape;
+    std::vector<InputShape> shapes;
+    ov::element::Type model_type;
+    std::string device_name;
+    std::tie(shapes, indices_shape, axis_batch_idx, model_type, device_name, indicesData, str_data) = basicParams;
```
Comment (t-jankowski) [inline src/tests/functional/shared_test_classes/src/single_op/gather.cpp:259]: `std::tie` here implies unnecessary copy of dynamically allocated containers which aren't then modified. `const &` per each tuple element is a better choice, although less convenient. This is a common practice in tests, so no need to change - just a comment.

File: src/plugins/template/tests/functional/op_reference/base_reference_test.hpp [hunk 5]
```diff
@@ -57,6 +57,11 @@ ov::Tensor CreateTensor(const ov::element::Type& element_type, const std::vector
 // Create blob with correct input shape (not 1-dimensional). Will be used in tests with dynamic input shapes
 template <class T>
 ov::Tensor CreateTensor(const ov::Shape& shape, const ov::element::Type& element_type, const std::vector<T>& values) {
+    if (element_type == ov::element::string) {
+        ov::Tensor tensor{element_type, shape};
+        std::copy_n(values.data(), shape_size(shape), tensor.data<T>());
+        return tensor;
+    }
```
Comment (t-jankowski) [inline src/plugins/template/tests/functional/op_reference/base_reference_test.hpp:64]: It might break RVO.

**Review by t-jankowski — APPROVED** [review]

File: src/plugins/template/tests/functional/op_reference/base_reference_test.hpp (hunk 5 above)
Comment (mitruska) [reply src/plugins/template/tests/functional/op_reference/base_reference_test.hpp:64]: Any suggestions for particular improvement?
This is a part of the test util corresponding to the existing code for the numeric types below:
https://github.com/openvinotoolkit/openvino/blob/c29bc760b176bce1202dc51261bfcd49eb459f88/src/plugins/template/tests/functional/op_reference/base_reference_test.hpp#L65-L72
Assuming the copy of the test data is done to ensure the life time, 
Is there a constructor of the ov::Tensor making a copy of the data from the std::vector?

Comment (praasz) [reply src/plugins/template/tests/functional/op_reference/base_reference_test.hpp:64]:
```cpp
    ov::Tensor tensor{element_type, shape};
    if (element_type == ov::element::string) {    
         std::copy_n(values.data(), shape_size(shape), tensor.data<T>());
    } else {
        size_t size = sizeof(T) * values.size();
        if (tensor.get_byte_size() < size)
            size = tensor.get_byte_size();
        std::memcpy(tensor.data(), values.data(), size);
    }
    return tensor;
```

1. Create tensor.
2. Depends on element type apply action.
3. return created tensor, RVO will be applied (always same object returned).

File: src/tests/functional/shared_test_classes/src/single_op/gather.cpp (hunk 3 above)
Comment (mitruska) [reply src/tests/functional/shared_test_classes/src/single_op/gather.cpp]: Done

File: src/tests/functional/shared_test_classes/src/single_op/gather.cpp (hunk 2 above)
Comment (mitruska) [reply src/tests/functional/shared_test_classes/src/single_op/gather.cpp]: Done

File: src/plugins/template/tests/functional/op_reference/base_reference_test.hpp (hunk 5 above)
Comment (praasz) [reply src/plugins/template/tests/functional/op_reference/base_reference_test.hpp]: I should be more correct the solution will apply NRVO.

**Review by mmikolajcz — APPROVED** [review]

**Pull request merged**